File: src/network/packet/RequestPacket.php
<?php

namespace pocketcloud\cloud\network\packet;

use pocketcloud\cloud\network\client\ServerClient;
use pocketcloud\cloud\network\packet\util\PacketData;

/**
 * The normal request packet sent from sub-servers to the cloud, which will answer through regular ResponsePacket
 * @see ResponsePacket
 */
abstract class RequestPacket extends CloudPacket implements CloudboundPacket {

    private string $requestId = "";

    final public function encode(PacketData $packetData): void {
        parent::encode($packetData);
        $packetData->write($this->requestId);
    }

    final public function decode(PacketData $packetData): void {
        parent::decode($packetData);
        $this->requestId = $packetData->readString();
    }

    public function sendResponse(ResponsePacket $packet, ServerClient $client): void {
        $client->sendPacket($packet->setRequestId($this->requestId));
    }

    public function getRequestId(): string {
        return $this->requestId;
    }
}

File: src/main/java/de/pocketcloud/cloud/network/packet/RequestPacket.java
package de.pocketcloud.cloud.network.packet;

import de.pocketcloud.cloud.network.client.ServerClient;
import de.pocketcloud.cloud.network.packet.data.PacketData;
import lombok.Getter;

/**
 * The normal request packet sent from sub-servers to the cloud, which will answer through a regular ResponsePacket.
 * @see ResponsePacket
 */
@Getter
public abstract class RequestPacket extends CloudPacket implements CloudboundPacket {

    private String requestId = "";

    @Override
    public final void encode(PacketData packetData) {
        super.encode(packetData);
        packetData.write(requestId);
    }

    @Override
    public final void decode(PacketData packetData) {
        super.decode(packetData);
        requestId = packetData.readString();
    }

    @Override
    public final void encodePayload(PacketData packetData) {}

    public void sendResponse(ResponsePacket packet, ServerClient client) {
        packet.setRequestId(requestId);
        client.channel().writeAndFlush(packet);
    }
}